chainntnfs/neutrinonotify/neutrino.go

```diff
@@ -631,6 +631,102 @@ func (n *NeutrinoNotifier) historicalConfDetails(confRequest chainntnfs.ConfRequ
 	return nil, nil
 }
 
+// historicalScriptSpend performs a historical spend scan when only a script
+// is provided. It fetches filters for every block height and checks to see
+// if the filter matches the provided PkScript. Unlike historicalConfDetails,
+// this function will scan forwards from startHeight towards endHeight instead
+// of the reverse.
+func (n *NeutrinoNotifier) historicalScriptSpend(
+	spendReq chainntnfs.SpendRequest, startHeight,
+	endHeight uint32) error {
+
+	// This variable holds our spendDetails if they're found and will be
+	// used to update the spending details of the SpendRequest parameter.
+	var spendDetails *chainntnfs.SpendDetail
+
+scanLoop:
+	for scanHeight := startHeight; scanHeight <= endHeight; scanHeight++ {
+		// If the notifier is shutting down, make sure we do too.
+		select {
+		case <-n.quit:
+			return chainntnfs.ErrChainNotifierShuttingDown
+		default:
+		}
+
+		// Fetch the block hash for the given scanHeight.
+		blockHash, err := n.p2pNode.GetBlockHash(int64(scanHeight))
+		if err != nil {
+			return fmt.Errorf("unable to get hash for height=%v:"+
+				"%v", scanHeight, err)
+		}
+
+		// Now we can fetch the basic filter for this height.
+		regFilter, err := n.p2pNode.GetCFilter(
+			*blockHash, wire.GCSFilterRegular,
+			neutrino.NumRetries(5),
+			neutrino.OptimisticReverseBatch(),
```

Comment:
We can't use this as we're scanning forwards, but scanning backwards is better as channels remain open for a longer period of time. We should consolidate the common logic between this method and `historicalConfDetails` (including the scanning portion) into a single method that we can use for both.

Reply:
For other reviewers: I'll remove `OptimisticReverseBatch`, but I think that forward scanning is the correct way as if we are notifying on a confirmation or spend without the txid, we could end up in a situation where we catch a later conf/spend, when we should catch the earliest possible

Reply:
I have removed this bit, need to check if it's safe to scan backwards if no outpoint is present in the confirmation case.  But that can be deferred in another PR.

Reply:
I think it depends on the caller of `RegisterConfirmationsNtfn` and their provided height hint.  But from my reading of the code, it seems that the backwards scanning in `historicalConfDetails` if only using a pkscript and no txid may be unsafe and notify on a *later* confirmation than desired.  The same logic exists in the `btcd` & `bitcoind` cases and I think it should be changed to do forwards scanning.